### ISSUE TITLE
French Accent encode munging

### ISSUE THREAD

**Issue by Sudburr** [notepad-plus-plus/notepad-plus-plus#14299]
### Steps to Reproduce the Issue

1. open notepad++ 8.5.8 (64bit) / new document
2. change language to Powershell or leave as .txt
3. paste Français
6. save file
7. exit notepad++
8. open the file in powershell ISE, notepad, a hex editor, and notepad++ 

### Expected Behavior

Français

### Actual Behavior

FranÃ§ais

### Debug Information

Notepad++ v8.5.8   (64-bit)
Build time : Oct 15 2023 - 21:43:56
Path : C:\Program Files\Notepad++\notepad++.exe
Command Line : "C:\temp\_input\Autodesk_Inventor_Professional_2024.1.1\export2.ps1"
Admin mode : OFF
Local Conf mode : OFF
Cloud Config : OFF
OS Name : Windows 11 Pro (64-bit)
OS Version : 23H2
OS Build : 22635.2552
Current ANSI codepage : 1252
Plugins : 
    mimeTools (2.9)
    NppConverter (4.5)
    NppExport (0.4)

... and ...

Notepad++ v8.5.8   (64-bit)
Build time : Oct 15 2023 - 21:43:56
Path : C:\Program Files\Notepad++\notepad++.exe
Command Line : 
Admin mode : OFF
Local Conf mode : OFF
Cloud Config : OFF
OS Name : Windows 10 Pro (64-bit)
OS Version : 22H2
OS Build : 19045.3570
Current ANSI codepage : 1252
Plugins : 
    mimeTools (2.9)
    NppConverter (4.5)
    NppExport (0.4)

**Comment by alankilborn** [comment]
When I try it, I get Français in Notepad++.
What is your encoding on the status bar?

**Comment by Sudburr** [comment]
Good morning.

Notepad++ will DISPLAY Français, but look at it using Powershell ISE and a HEX editor to see its true form.

= = = = =
***@***.***
ECNO Shared Technology Services Analyst
[cid:23a780d6-a301-4aa2-8524-9ba40b65d646]

________________________________

**Comment by alankilborn** [comment]
> ... true form

Notepad++ not only _displays_ it correctly, but for UTF-8 encoding, the underlying bytes when viewed in a hex editor _are also correct_; ref.: https://www.compart.com/en/unicode/U+00E7 to see the character and its byte representation.

You didn't say what encoding you are using (depite my direct question before in this regard), so I can only assume UTF-8...

It might help if you stated what you _expect_ for the bytes that you're seeing in a hex editor, rather than just saying "it's munged" (per the title of this issue).

---

BTW, notepad.exe, like Notepad++, for me has no problem with display:

![image](https://github.com/notepad-plus-plus/notepad-plus-plus/assets/77065706/38ce71f3-d804-4e18-9b87-b2061e05cf2f)

**Comment by Sudburr** [comment]
![image](https://github.com/notepad-plus-plus/notepad-plus-plus/assets/149632684/8acba969-1724-4fe9-a153-7605b903afeb)
UTF-8 . Default install.

I have reproduced on Win11 and Win10.

I found it while using a .ps1 to search for a value in the registry. It looks fine in Notepad++, and Notepad actually, but looking at the same file in the Powershell ISE or using HxD ( https://mh-nexus.de/en/hxd/ ) it's clearly broken.

In the following shot I created the file "new 1.txt", then opened that in notepad, then opened the file in powershell, then opened the file using HxD in Windows 10, then opened the file in powershell ISE in Windows 10.
![image](https://github.com/notepad-plus-plus/notepad-plus-plus/assets/149632684/200af595-cf65-4abf-9c4c-df5fdf271942)

**Comment by Sudburr** [comment]
Language and region settings of OS
![image](https://github.com/notepad-plus-plus/notepad-plus-plus/assets/149632684/068f7733-ae14-424c-9758-a3b9e357b116)

**Comment by Sudburr** [comment]
"Français" was copied from the registry after installing Autodesk AutoCAD 2024.1. Then pasted into the script while I built it in Notepad++ . For today's example, I used the Français that I included in the original ticket text on this webpage.

**Comment by Sudburr** [comment]
Export of original registry attached.
[export.zip](https://github.com/notepad-plus-plus/notepad-plus-plus/files/13242108/export.zip)

**Comment by pryrt** [comment]
You have a file whose bytes are saved as UTF8. Those are the correct bytes.  Notepad++ knows that the file is UTF-8, and reads those two bytes as the single character, which is the correct behavior.

> then opened that file using HxD in the Windows 10 VM = FranÃ§ais

As is obvious from the "Windows (ANSI)" in your screenshot, HxD is interpreting those bytes as ANSI, not as UTF-8.  
![image](https://github.com/notepad-plus-plus/notepad-plus-plus/assets/17455758/2d9de316-dda6-40eb-a3a9-3343fade4b65)

Hence, it is misinterpreting them as the two ANSI characters `Ã` and `§`, instead of the single UTF-8 character `ç` .  

Unfortunately for you, HxD does not appear to give the option of UTF-8 in that pulldown.  Fortunately for you, if you move the cursor to the first byte (what appears as `Ã` in the `Ã§`), and scroll on the right until you see UTF-8 code point, you will see that it does know it's the `ç` -- it just doesn't display that in the its editor panel (because it's showing you individual bytes interpreted as ANSI).

So this "discrepancy" in HxD was just you not understanding that HxD will _never_ show UTF-8 characters correctly in the main editor panel, because that's not how it was designed.  Notepad++ correctly wrote the `ç`

![image](https://github.com/notepad-plus-plus/notepad-plus-plus/assets/17455758/c0702e98-38d7-444c-97e8-85e5ca0c15e4)

> 9. then opened that file in powershell ISE in Windows 10 = FranÃ§ais

I'm not a PowerShell ISE user, so I wasn't sure if it had a way to tell it to read UTF-8 vs ANSI.  But I just did a quick google for you, and found that PowerShell ISE will only interpret a file as UTF-8 if it has the BOM indicator at the start of the file.  

You didn't tell Notepad++ to use UTF-8-BOM, so it's not writing the BOM character, so PowerShell ISE incorrectly assumes it's ANSI-encoded.  If you want Notepad++ to save it with the BOM, use Notepad++ > **Encoding > UTF-8-BOM** and then save.  Once you do that, then open the file in PowerShell ISE , whcih will then interpret the file as UTF-8 like you want it to.

Here is a screenshot of PowerShell ISE showing the non-BOM UTF-8 file as ANSI, plus I used the copy of [xxd.exe](https://sourceforge.net/projects/xxd-for-windows/) that I have to show the raw hex bytes for both the UTF-8 (`francais.txt`) and UTF-8-BOM (`francais_bom.txt`) files:
![image](https://github.com/notepad-plus-plus/notepad-plus-plus/assets/17455758/154b6678-36a4-4c6c-8251-76c7db15922a)

And then when I switch to `francais_bom.txt` in PowerShell ISE, it sees the BOM and correctly
![image](https://github.com/notepad-plus-plus/notepad-plus-plus/assets/17455758/c7e70b49-018d-48c7-9424-d954d88c3edb).

So Notepad++ is correctly writing a UTF-8 file.  And if you tell it to use a BOM, then PowerShell ISE will read it as a UTF-8 file.

If you don't want to use UTF-8, the default ANSI Windows-1252 encoding has a codepoint for `ç` as well, so if you **Convert to ANSI** in Notepad++ and save it to `francais_ansi.txt`, then look at it in xxd and PowerShell ISE, it also works correctly.

![image](https://github.com/notepad-plus-plus/notepad-plus-plus/assets/17455758/b30af21b-f5b3-41af-847d-ef0861d79cc5)

To sum up: Notepad++ was correctly doing what you told it to do.  You misunderstood what HxD was telling you.  And PowerShell ISE has design limitations that means you either need the file saved with a BOM for it to be UTF-8, or you need the file saved in ANSI (Windows-1252 on your machine) to be able to _show_ you a correctly-encoded  `ç` .  So whether you use Notepad++ to add the BOM to a UTF-8 file, or use Notepad++ to convert the file to ANSI and save it, either way, Notepad++ can and will write the file in the format that PowerShell ISE desires: you just have to know how to accommodate the limitations of that Microsoft tool.

**Comment by alankilborn** [comment]
@Sudburr The bottom line is that you reported a problem with Notepad++, it was investigated and determined that there is no problem with Notepad++ in this regard, and thus this issue can and should be closed (so, as the petitioner, please close it -- thank you).

**Comment by Sudburr** [comment]
Interesting ... and I think I've found the source of my angst.

When I start creating a new .ps1 using Notepad++, the default new file is a .txt with UTF-8 encoding. When I switch the language to Powershell. the encoding does not follow automatically to UTF-8-BOM. When I switch to Registry the encoding does not automatically switch to UTF-16 LE BOM.

So the scripts I've created from scratch using Notepad++ don't have the correct encoding unless I change it manually before saving it.

I've not made any changes to Notepad++. Does that option exist?

**Comment by pryrt** [comment]
> When I switch the language to Powershell. the encoding does not follow automatically to UTF-8-BOM. When I switch to Registry the encoding does not automatically switch to UTF-16 LE BOM. ... Does that option exist?

Notepad++ has **Settings > Preferences > New Document > Encoding**, which allows you to pick the encoding for all new documents.  But it doesn't have an "automatically change my encoding if I save it as a particular file type" option, which is what you seem to want.

(And as I said above, PowerShell also accepts ANSI encoded files.  And Registry files can be ANSI-encoded, too.  So there is nothing about those tools that _requires_ a particular encoding; it's just that PowerShell ISE editor won't recognize the file as UTF-8 unless it has the BOM.  I don't know what Windows would do if you ran a .reg file that was UTF-8 encoded, with or without a BOM, but I expect that it would at least work right with a BOM.)

**Comment by Sudburr** [comment]
To play nice in Microsoft's sandbox of Windows, shouldn't it automatically adjust encoding to the standard (assuming that MS has published it as a standard) that Windows uses to ensure compliance? Perhaps a warning when changing languages of what is being worked on?

**Comment by pryrt** [comment]
> To play nice in Microsoft's sandbox of Windows, shouldn't it automatically adjust encoding to the standard (assuming that MS has published it as a standard) that Windows uses to ensure compliance? 

Notepad++ cannot know all the encoding quirks for all the 80+ different filetypes that it can save by default (and the essentially infinite other text-based formats that aren't listed in the lexers).

> Perhaps a warning when changing languages of what is being worked on?

That warning would be a lie, because UTF-8-BOM is not the only valid encoding for PS1 files.  UTF-8-BOM is _not_ a requirement of PowerShell scripts.  It is _one_ of the valid encodings for the outdated PowerShell ISE application (ANSI or UTF-8-BOM).  It is not a limitation of `.ps1` that ISE cannot read a UTF-8 script unless it includes the BOM.  `.ps1` scripts also work in UTF-16-LE and UTF-16-BE (both of which always have a BOM), and of course in ANSI as well... so there are many valid encodings for `.ps1` files.

Notepad++ expects its users to understand the encoding needs of any other software/application that needs to be able to read the files that it produces, and for the user to set up Notepad++ appropriately to produce that kind of file.

Also, according to my research, [PowerShell ISE is outdated](https://stackoverflow.com/questions/57131654/using-utf-8-encoding-chcp-65001-in-command-prompt-windows-powershell-window/57134096#57134096:~:text=Why%20using%20the%20Windows%20PowerShell%20ISE%20is%20ill%2Dadvised%20in%20general) and is apparently not a good idea for PowerShell development, even outside of its inability to recognize a valid encoding.  Honestly, you should be using Notepad++'s **Run** menu, which could be used to launch the active file with Power Shell into a PowerShell window, rather than going to the intermediate step of the ISE.

**Comment by Sudburr** [comment]
Thank you for the links. Sage advice. It appears I may need to upgrade my methodology to use VS Code now ... and watch my encoding when in Notepad++.

**Comment by Sudburr** [comment]
Thank you for the links. Sage advice. It appears I may need to upgrade my methodology to use VS Code now ... and watch my encoding when in Notepad++.
